h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/matrix/SparseMatrixFactory.java

```diff
@@ -49,15 +51,68 @@ public static DMatrix csr(
             chunks, weight,
             di, sparseMatrix._rowHeaders, sparseMatrix._sparseData, sparseMatrix._colIndices,
             respIdx, resp, weights, offsetIdx, offsets);
-        return toDMatrix(sparseMatrix, sparseMatrixDimensions, actualRows, di);
+        return toDMatrix(sparseMatrix, sparseMatrixDimensions, actualRows, di.fullN(), resp, weights, offsets).get();
+    }
+    
+    public static class SparseDMatrixProvider extends MatrixLoader.DMatrixProvider implements Serializable {
+
+        private final long[][] rowHeaders;
+        private final int[][] colIndices;
+        private final float[][] sparseData;
+        private final DMatrix.SparseType csr;
+        private final int shape;
+        private final long nonZeroElementsCount;
+
+        public SparseDMatrixProvider(
+            long[][] rowHeaders,
+            int[][] colIndices,
+            float[][] sparseData,
+            DMatrix.SparseType csr,
+            int shape,
+            long nonZeroElementsCount,
+            int actualRows,
+            float[] response,
+            float[] weights,
+            float[] offsets
+        ) {
+            super(actualRows, response, weights, offsets);
+            this.rowHeaders = rowHeaders;
+            this.colIndices = colIndices;
+            this.sparseData = sparseData;
+            this.csr = csr;
+            this.shape = shape;
+            this.nonZeroElementsCount = nonZeroElementsCount;
+        }
+
+        @Override
+        public DMatrix makeDMatrix() throws XGBoostError {
+            return new DMatrix(rowHeaders, colIndices, sparseData, csr, shape, (int) actualRows+1, nonZeroElementsCount);
+        }
+
+        @Override
+        public void writeTo(DataOutputStream os) throws IOException {
+            os.write(SPARSE);
+            try (ObjectOutputStream out = new ObjectOutputStream(os)) {
+                // TODO find out if a custom serializer has any benefits over java serialization
```

Comment:
need to do some experimenting whether (de)serialisaion via direct usage of dataOutputStream would be faster/more efficient than java serialisation